File: lc/coreprep/implementation/practice_iterations/topological_sort_2/cycle_directed_graph_dfs.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

bool dfs(unordered_map<int, unordered_set<int>> &graph, vector<bool> &visited,
         vector<bool> &call_st, int curr_vertex) {
  if (call_st[curr_vertex]) {
    return true;
  }

  visited[curr_vertex] = true;
  call_st[curr_vertex] = true;

  for (int neighbour : graph[curr_vertex]) {
    if (dfs(graph, visited, call_st, neighbour)) {
      return true;
    }
  }

  call_st[curr_vertex] = false;
  return false;
}

// O(V + E) / O(V + E)
bool has_cycle(unordered_map<int, unordered_set<int>> &graph) {
  vector<bool> visited(graph.size());
  vector<bool> call_st(graph.size());

  for (int i = 0; i < graph.size(); i++) {
    if (!visited[i]) {
      if (dfs(graph, visited, call_st, i)) {
        return true;
      }
    }
  }
  return false;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m, from, to;
  cin >> tt;

  while (tt--) {
    cin >> n;

    unordered_map<int, unordered_set<int>> graph;
    for (int i = 0; i < n; i++) {
      graph[i] = unordered_set<int>{};
    }

    cin >> m;
    for (int i = 0; i < m; i++) {
      cin >> from >> to;
      graph[from].insert(to);
    }

    cout << has_cycle(graph) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/CycleDirectedGraphDFS.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class CycleDirectedGraphDFS {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) {
          graph.put(i, new HashSet<>());
        }

        int m = in.nextInt();
        for (int i = 0; i < m; i++) {
          int from = in.nextInt();
          int to = in.nextInt();
          graph.get(from).add(to);
        }

        System.out.println(hasCycle(graph));
      }
    }
  }

  // O(V + E) / O(V + E)
  public static boolean hasCycle(Map<Integer, Set<Integer>> graph) {
    boolean[] visited = new boolean[graph.size()];
    boolean[] callSt = new boolean[graph.size()];

    for (int i = 0; i < graph.size(); i++) {
      if (!visited[i]) {
        if (dfs(graph, visited, callSt, i)) {
          return true;
        }
      }
    }
    return false;
  }

  private static boolean dfs(Map<Integer, Set<Integer>> graph,
                             boolean[] visited, boolean[] callSt,
                             int currVertex) {
    if (callSt[currVertex]) {
      return true;
    }

    visited[currVertex] = true;
    callSt[currVertex] = true;

    for (int neighbour : graph.get(currVertex)) {
      if (dfs(graph, visited, callSt, neighbour)) {
        return true;
      }
    }

    callSt[currVertex] = false;
    return false;
  }
}